### DISCUSSION TITLE
How can I run mice with a GAMM?

### DISCUSSION THREAD

**Discussion by kjetil-uis** [amices/mice#596] [Getting started] [answered]
I try to impute a GAMM model but can not figure out how. 
If I use `predictorMatrix` I can set the class variable for random effect, but no splines or variable crossing.
If I use `formulas can set splines and variable crossings but I am not able to set the random/class variable. 

I tried this: 

``` 
imp  <- mice(nhanes,
                       formulas = list( 
                            #bmi = bmi ~ s(chl) + (1 + age | hyp)
                            bmi = bmi ~ poly(chl) + (1 + age | hyp),
                            chl = chl ~ age + bmi + hyp,
                            hyp = hyp ~ bmi + chl*poly(age)
                       ), 
                       #method = "norm.predict", # runs when s() is not present 
                       #method = "2l.norm",         # Error: No class variable 
                       #method = "2l.pan",           # Error: No class variable 
                       #method = "2l.lmer",          # Error: lmer does not run 
                       #method = '2l.pmm',          # Error in str2lang(x) : <text>:1:27: unexpected ')'
                       #method = "2l.glm.norm",   # Error in factor(x[, clust], labels = 1:n.class) :invalid 'labels'; length 2 should be 1 or 0
                       print = FALSE
                 )
```

When using method = ´norm.predict´ it runs with `poly()` instead of `s()` - but does it account for the random effect?
If I use method = '2l.method' it does not run (Error is commented in the code above). 

What is wrong?

**Comment by stefvanbuuren** [comment, marked as answer]
You are trying to specify imputation models that are more complex than `mice()` can handle. Generalised additive mixed models (GAMM) are not supported. The best advice that I can offer at the moment is to simplify your model in one of two ways:

1. run a flat model (no class variable) with data transforms. This may be sensible if the ICC is low (so clustering is not that important) and data transforms are critical;
2. run a linear multilevel model (no transforms) if ICC is high, and use an imputation method that is relatively robust against violations of linearity, e.g., `2l.pmm()`.

Your choice would depend on the data at hand.

**Reply by kjetil-uis** [reply]
Thank you for a quick response. 
I have longitudinal data with missing in only the dependent variable, so I think alternative 2 would be most appropriate.  

To include non-linearity is not critical, but I would like to have random slope. Is it possible?  By using `formulas` notation like `(1 + age | hyp)` or other ways?

**Reply by stefvanbuuren** [reply]
You should be able to find that in the documentation.

**Reply by stefvanbuuren** [reply]
Addition: This is work in progress. 

See <https://github.com/amices/mice/pull/582/commits/f5d5c99a325a3f16875925ba7e35417d75ed2fbe> line 624: 
>> The random slope version `fml <- list(ReadDis ~ ReadAchiev + (1 + ReadAchieve|ID))` does not yet work due to improper formula processing by `mice()`

But having a random slope and intercept is possible by `predictorMatrix` route.

**Reply by kjetil-uis** [reply]
It will be great when you can use `formulas` - notation.

I found that for `2l.pan`you can use 4 in the `predictorMatrix` to get a random slope. 

Also I was also able to run `formulas = list( y ~ poly(x,2) + (1 + x | ID))` with `method = "jomoImpute"`. However I was not able to plot the convergence -plot and it took 10min vs 0.1 min when I used `mitml:jomoImpute` directly on the same data - so I am a bit uncertain how much trust I should put in that result. 

Thank you again and keep up the good work.
